### PR TITLE
Support pg_tablespace_size() for non-default tablespaces

### PR DESCRIPTION
**Pull request by dgkimura**
Issue is that calculate_tablespace_size() has an expectation that the
directory to get the size of can be accessed as a relative path from
inside data directory. Since filespaces are not necessarily in the data
directory, a simple solution is to create a symlink.

### CODE REVIEW

File: src/backend/storage/smgr/md.c [hunk 1]
```diff
@@ -617,6 +618,12 @@ mdcreatetablespacedir(
 
 		if (mkdir(tablespacePath, 0700) < 0)
 			*primaryError = errno;
+
+		location = (char *) palloc(10 + 10 + 1);
+		sprintf(location, "pg_tblspc/%u", tablespaceOid);
+		if (symlink(tablespacePath, location))
```
Comment (asimrp) [inline src/backend/storage/smgr/md.c]: How does this symlink get deleted?  Any implications on backup / restore (`gpcrondump` is the backup utility in 5X)?

Comment (soumyadeep2007) [reply src/backend/storage/smgr/md.c]: Perhaps we could use `psprintf` here.

Comment (dgkimura) [reply src/backend/storage/smgr/md.c]: Thanks Asim! I had a strong feeling I was missing something.  Next revision deletes the symlink in `mdrmtablespacedir()`.  However, I still need to look into `gpcrondump`.

Thanks for suggestion Deep!  I think to make it even simpler I'll just make the variable a char array on the stack. Then no palloc necessary.

**Comment by ashwinstar** [comment]
I need to give detailed thought to this. I feel extremely reserved in making this kind of on-disk layout change in stable release. If really pressing necessity then we should do it else seems not worth the complexity. As we will need to deal in new code, if the link exist or not, as it will not for existing installations. So, for those cases anyways the function will continue to fail.

Also, given tablespace is really thin abstraction on top of filespace in 5X. Filespace is really the holder of actually data on-disk. So, getting size of tablespace is what was really intended or getting filespace size?

**Comment by Eulerizeit** [comment]
On Fri, Sep 27, 2019 at 8:59 AM Ashwin Agrawal <notifications@github.com>
wrote:

> I need to give detailed thought to this. I feel extremely reserved in
> making this kind of on-disk layout change in stable release. If really
> pressing necessity then we should do it else seems not worth the
> complexity. As we will need to deal in new code, if the link exist or not,
> as it will not for existing installations. So, for those cases anyways the
> function will continue to fail.
>
> Also, given tablespace is really thin abstraction on top of filespace in
> 5X. Filespace is really the holder of actually data on-disk. So, getting
> size of tablespace is what was really intended or getting filespace size?
>

If there is significant risk in this I'm ok with NOT doing this.
Effectively it is a feature that benefits some users but the value IMO is
not worth potential instability.

-- Rob

**Comment by soumyadeep2007** [comment]
The way I see it is that end users may not only want to see what each filespace consumes as disk space but also see how much each tablespace consumes within each filespace.

I agree that we shouldn't introduce a symlink to keep `calculate_tablespace_size()` happy. Instead I would do the following:

1. Have calculate_tablespace_size modify this branch :
```
else
	snprintf(tblspcPath, MAXPGPATH, "pg_tblspc/%u", tblspcOid);
``` 
a) Query `pg_filespace_entry` to obtain the entry corresponding to the current dbid, specifically the `fselocation`. 
b) Append the tablespace's oid to the fselocation to obtain the tablespace's path.

After that the rest of `calculate_tablespace_size` should just work.

**Comment by dgkimura** [comment]
On Mon, Sep 30, 2019 at 10:45 AM Soumyadeep Chakraborty <notifications@github.com> wrote:
> Have calculate_tablespace_size modify this branch :
>
> else
> snprintf(tblspcPath, MAXPGPATH, "pg_tblspc/%u", tblspcOid);
>
> a) Query pg_filespace_entry to obtain the entry corresponding to the current dbid, specifically the fselocation.
> b) Append the tablespace's oid to the fselocation to obtain the tablespace's path.
>
> After that the rest of calculate_tablespace_size should just work.

Thanks Deep!  I think that sounds reasonable.  I'll take a look.

**Comment by berlin-ab** [comment]
Nice suggestion @soumyadeep2007 .

**Comment by asimrp** [comment]
Closing this in favor of https://github.com/greenplum-db/gpdb/pull/8863, that implements Deep's suggestion.

**Comment by asimrp** [comment]
@dgkimura instead of force-pushing on your branch, I chose to open another PR.

**Pull request closed without merging**